File: py_algorithms/Programmers/heap_01_more_spicy.py
# Process
# 1. Input scoville array and threshold K
# 2. Make heap queue
# 3. Iterate while all scoville is over K
#  3.1. Check scoville (smallest one)
#   3.1.1. If passed -> done.
#   3.1.2. If not -> mix foods
#  3.3. Count
# 4. Return answer

import heapq

def solution(scoville, K):
    count = 0
    
    # 2. 
    heap_scoville = []
    for value in scoville:
        heapq.heappush(heap_scoville, value)
    
    # 3.
    is_done = False
    while len(heap_scoville) >= 1:
        first = heapq.heappop(heap_scoville)
        if first >= K: 
            is_done = True
            break
        else:
            if len(heap_scoville) > 0:
                second = heapq.heappop(heap_scoville)
                mixed = mix_food(first, second)
                heapq.heappush(heap_scoville, mixed)
        count += 1
        
    if not is_done:
        return -1
        
    return count



def mix_food(food_a, food_b):
    if food_a > food_b:
        return food_b + food_a * 2
    else:
        return food_a + food_b * 2

File: src/main/java/algorithm_sites/programmers/Heap_01_MoreSpicy.java
package algorithm_sites.programmers;

import java.util.PriorityQueue;
import java.util.Queue;

public class Heap_01_MoreSpicy {

    public int solution(int[] scoville, int K) {
        Queue<Long> priorityQueue = new PriorityQueue<>();

        for (int point : scoville) {
//            if (point < K)
                priorityQueue.add((long) point);
        }

        int count = 0;

        boolean isDone = false;
        while (!isDone && priorityQueue.size() > 1) {
            Long firstMin = priorityQueue.poll();
            Long secondMin = priorityQueue.poll();

            if (firstMin >= K && secondMin >= K) {
                isDone = true;
                break;
            }

            ++count;

            long mix = mix(firstMin, secondMin);

            priorityQueue.add(mix);
        }

        if (priorityQueue.size() == 0) {
            return count;
        }

        if (priorityQueue.size() == 1 && priorityQueue.peek() >= K) {
            return count;
        }

        if (isDone) {
            return count;
        }

        return -1;
    }

    private long mix(long first, long second) {
        return first + (second * 2);
    }

}
